### PR TITLE
Implemented SPV_INTEL_runtime_aligned extension

### PR DESCRIPTION
**Pull request by MrSidims**
This extension introduces a new function parameter attribute that can be
applied to pointers to indicate that the pointer was allocated with an
implementation-specific alignment.

Spec:
https://github.com/KhronosGroup/SPIRV-Registry/blob/c7079ba821f5ec70332ad6ec3a4962bfcd7f17ac/extensions/INTEL/SPV_INTEL_runtime_aligned.asciidoc

Signed-off-by: Leonid Pauzin <leonid.pauzin@intel.com>
Signed-off-by: Dmitry Sidorov <dmitry.sidorov@intel.com>

### CODE REVIEW

**Comment by MrSidims** [comment]
@AGindinson @mlychkov please review

File: lib/SPIRV/SPIRVWriter.cpp [hunk 1]
```diff
@@ -611,6 +611,12 @@ SPIRVFunction *LLVMToSPIRVBase::transFunctionDecl(Function *F) {
   if (BM->isAllowedToUseExtension(ExtensionID::SPV_INTEL_fpga_buffer_location))
     BufferLocation = ((*F).getMetadata("kernel_arg_buffer_location"));
 
+  // Translate runtime_aligned metadata if it's attached to the translated
+  // function declaration
+  MDNode *RuntimeAligned = nullptr;
+  if (BM->isAllowedToUseExtension(ExtensionID::SPV_INTEL_runtime_aligned))
+    RuntimeAligned = ((*F).getMetadata("kernel_arg_runtime_aligned"));
```
Comment (AGindinson) [inline lib/SPIRV/SPIRVWriter.cpp:618]: Wouldn't this still work? Or is some type magic necessary here?
```suggestion
    RuntimeAligned = F->getMetadata("kernel_arg_runtime_aligned");
```

File: lib/SPIRV/SPIRVWriter.cpp [hunk 2]
```diff
@@ -649,6 +655,17 @@ SPIRVFunction *LLVMToSPIRVBase::transFunctionDecl(Function *F) {
...
     }
+    if (RuntimeAligned && I->getType()->isPointerTy()) {
+      // Order of integer numbers in MD node follows the order of function
+      // parameters on which we shall attach the appropriate decoration. Add
+      // decoration only if MD value is 1.
+      int LocID = 0;
+      if (!isa<MDString>(RuntimeAligned->getOperand(ArgNo)) &&
+          !isa<MDNode>(RuntimeAligned->getOperand(ArgNo)))
+        LocID = getMDOperandAsInt(RuntimeAligned, ArgNo);
+      if (LocID > 0)
```
Comment (AGindinson) [inline lib/SPIRV/SPIRVWriter.cpp:666]:
```suggestion
      if (LocID == 1)
```
...seems more cohesive with the above comment:
> Add decoration only if MD value is 1

File: lib/SPIRV/SPIRVReader.cpp [hunk 3]
```diff
@@ -191,6 +191,27 @@ static void addBufferLocationMetadata(
                     MDNode::get(*Context, ValueVec));
 }
 
+static void addRuntimeAlignedMetadata(
+    LLVMContext *Context, SPIRVFunction *BF, llvm::Function *Fn,
+    std::function<Metadata *(SPIRVFunctionParameter *)> Func) {
```
[the remark this answers is not in the document]
Comment (MrSidims) [reply lib/SPIRV/SPIRVReader.cpp:196]: Agree. If you don't mind I've renamed other Func parameters in the same PR.

**Review by AGindinson — APPROVED** [review]

**Pull request merged**